### PR TITLE
Add basic support for agent events

### PR DESCRIPTION
**Pull request by tklauser**
Support for agent events was added to the Hubble API in cilium/cilium#14168
    
Allow to print these messages in the Hubble CLI.

Note that this needs cilium/cilium#14242 to work  properly.

Example output (`hubble observe -t agent`):

```
% ./hubble observe -t agent           
TIMESTAMP             NODE                 TYPE                          DETAILS
Dec  2 13:53:15.272   kind-worker          ENDPOINT_CREATED              id: 2172, pod name: coredns-f9fd979d6-p4cz2, namespace: kube-system
Dec  2 13:53:15.298   kind-worker          IPCACHE_UPSERTED              cidr: 10.244.1.155/32, identity: 3, host ip: 172.18.0.2, encrypt key: 0
Dec  2 13:53:15.299   kind-control-plane   IPCACHE_UPSERTED              cidr: 10.244.1.155/32, identity: 3, host ip: 172.18.0.2, encrypt key: 0
Dec  2 13:53:15.633   kind-control-plane   ENDPOINT_CREATED              id: 793
Dec  2 13:53:16.093   kind-worker          ENDPOINT_CREATED              id: 2068
Dec  2 13:53:16.268   kind-worker          ENDPOINT_CREATED              id: 2456, pod name: hubble-relay-5d86f9dc74-np79w, namespace: kube-system
Dec  2 13:53:16.287   kind-control-plane   IPCACHE_UPSERTED              cidr: 10.244.1.105/32, identity: 3, host ip: 172.18.0.2, encrypt key: 0
Dec  2 13:53:16.287   kind-worker          IPCACHE_UPSERTED              cidr: 10.244.1.105/32, identity: 3, host ip: 172.18.0.2, encrypt key: 0
Dec  2 13:53:19.048   kind-control-plane   ENDPOINT_REGENERATE_SUCCESS   id: 2440, labels: [reserved:host,k8s:node-role.kubernetes.io/master]
Dec  2 13:53:19.057   kind-control-plane   ENDPOINT_REGENERATE_SUCCESS   id: 2440, labels: [reserved:host,k8s:node-role.kubernetes.io/master]
Dec  2 13:53:19.648   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 464, labels: [reserved:host]
Dec  2 13:53:19.649   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 464, labels: [reserved:host]
Dec  2 13:53:19.764   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2456, labels: [k8s:io.kubernetes.pod.namespace=kube-system,k8s:io.cilium.k8s.policy.serviceaccount=hubble-relay,k8s:io.cilium.k8s.policy.cluster=default,k8s:k8s-app=hubble-relay]
Dec  2 13:53:19.767   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2456, labels: [k8s:io.cilium.k8s.policy.cluster=default,k8s:k8s-app=hubble-relay,k8s:io.kubernetes.pod.namespace=kube-system,k8s:io.cilium.k8s.policy.serviceaccount=hubble-relay]
Dec  2 13:53:19.808   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2172, labels: [k8s:io.cilium.k8s.policy.cluster=default,k8s:k8s-app=kube-dns,k8s:io.kubernetes.pod.namespace=kube-system,k8s:io.cilium.k8s.policy.serviceaccount=coredns]
Dec  2 13:53:19.810   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2172, labels: [k8s:io.cilium.k8s.policy.serviceaccount=coredns,k8s:io.cilium.k8s.policy.cluster=default,k8s:k8s-app=kube-dns,k8s:io.kubernetes.pod.namespace=kube-system]
Dec  2 13:53:19.846   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2068, labels: [reserved:health]
Dec  2 13:53:19.847   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2068, labels: [reserved:health]
Dec  2 13:53:19.991   kind-control-plane   ENDPOINT_REGENERATE_SUCCESS   id: 793, labels: [reserved:health]
Dec  2 13:53:19.995   kind-control-plane   ENDPOINT_REGENERATE_SUCCESS   id: 793, labels: [reserved:health]
Dec  2 13:53:20.358   kind-worker          ENDPOINT_CREATED              id: 3165, pod name: local-path-provisioner-78776bfc44-vrbr5, namespace: local-path-storage
Dec  2 13:53:20.361   kind-control-plane   ENDPOINT_CREATED              id: 2672, pod name: coredns-f9fd979d6-l46xc, namespace: kube-system
Dec  2 13:53:20.368   kind-control-plane   ENDPOINT_REGENERATE_SUCCESS   id: 2440, labels: [k8s:node-role.kubernetes.io/master,reserved:host]
Dec  2 13:53:20.369   kind-control-plane   ENDPOINT_REGENERATE_SUCCESS   id: 793, labels: [reserved:health]
Dec  2 13:53:20.370   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2172, labels: [k8s:k8s-app=kube-dns,k8s:io.kubernetes.pod.namespace=kube-system,k8s:io.cilium.k8s.policy.serviceaccount=coredns,k8s:io.cilium.k8s.policy.cluster=default]
Dec  2 13:53:20.371   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2068, labels: [reserved:health]
Dec  2 13:53:20.372   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 2456, labels: [k8s:io.cilium.k8s.policy.cluster=default,k8s:k8s-app=hubble-relay,k8s:io.kubernetes.pod.namespace=kube-system,k8s:io.cilium.k8s.policy.serviceaccount=hubble-relay]
Dec  2 13:53:20.373   kind-worker          ENDPOINT_REGENERATE_SUCCESS   id: 464, labels: [reserved:host]
Dec  2 13:53:20.408   kind-worker          IPCACHE_UPSERTED              cidr: 10.244.0.29/32, identity: 3, host ip: 172.18.0.3, encrypt key: 0
Dec  2 13:53:20.409   kind-control-plane   IPCACHE_UPSERTED              cidr: 10.244.0.29/32, identity: 3, host ip: 172.18.0.3, encrypt key: 0
Dec  2 13:53:20.412   kind-worker          IPCACHE_UPSERTED              cidr: 10.244.1.159/32, identity: 3, host ip: 172.18.0.2, encrypt key: 0
Dec  2 13:53:20.412   kind-control-plane   IPCACHE_UPSERTED              cidr: 10.244.1.159/32, identity: 3, host ip: 172.18.0.2, encrypt key: 0
```

### CODE REVIEW

File: pkg/printer/printer.go
Comment (kAworu) [inline pkg/printer/printer.go:483]: Any reason to write to the error output here (i.e use `p.opts.werr` over `p.opts.w`)? It made my `hubble observe -t agent | jq` quite sad.

File: pkg/printer/printer.go
Comment (kAworu) [inline pkg/printer/printer.go:485]: Are you intending to address this in a future PR?

**Review by kAworu — CHANGES_REQUESTED** [review]
Awsome work @tklauser! I have a couple of comment about the output, but overall great PR.

File: pkg/printer/printer.go
Comment (tklauser) [reply pkg/printer/printer.go:485]: Oops, no I intended to address this already in this PR. Will implement as well.

File: pkg/printer/printer.go
Comment (tklauser) [reply pkg/printer/printer.go:483]: I guess I must have copy&pasted this from `WriteProtoNodeStatusEvent` above without thinking much. In fact I think we could just use `return p.jsonEncoder.Encode(r)` here. Which makes me realize that we probably want to distinguish `JSONOuutput` and `JSONPBOutput`. How about the following?

```go
        switch p.opts.output {
        case JSONOutput:
                return p.jsonEncoder.Encode(e)
        case JSONPBOutput:
                return p.jsonEncoder.Encode(r)
        // ...
```

where `e` is the `observerpb.AgentEvent` and `r` is the `observerpb.GetFlowsResponse`.

File: pkg/printer/printer.go
Comment (tklauser) [inline pkg/printer/printer.go:499]: I also realized that for the other event types we only write out the node name when `p.opts.nodeName` is set. Will adjust this accordingly as well.

**Comment by tklauser** [comment]
Addressed all of @kAworu's review comments, PTAL.

<details>
<summary>incremental diff</summary>

  ```diff
diff --git a/pkg/printer/printer.go b/pkg/printer/printer.go
index 9224288a52be..a8ee03e58642 100644
--- a/pkg/printer/printer.go
+++ b/pkg/printer/printer.go
@@ -479,24 +479,45 @@ func (p *Printer) WriteProtoAgentEvent(r *observerpb.GetFlowsResponse) error {
 	}
 
 	switch p.opts.output {
-	case JSONOutput, JSONPBOutput:
-		return json.NewEncoder(p.opts.werr).Encode(r)
+	case JSONOutput:
+		return p.jsonEncoder.Encode(e)
+	case JSONPBOutput:
+		return p.jsonEncoder.Encode(r)
 	case DictOutput:
-		// TODO
+		ew := &errWriter{w: p.opts.w}
+
+		if p.line != 0 {
+			ew.write(dictSeparator)
+		}
+
+		ew.write("  TIMESTAMP: ", fmtTimestamp(r.GetTime()), newline)
+		if p.opts.nodeName {
+			ew.write("       NODE: ", r.GetNodeName(), newline)
+		}
+		ew.write(
+			"       TYPE: ", e.GetType(), newline,
+			"    DETAILS: ", getAgentEventDetails(e), newline,
+		)
+		if ew.err != nil {
+			return fmt.Errorf("failed to write out agent event: %v", ew.err)
+		}
 	case TabOutput:
 		ew := &errWriter{w: p.tw}
 		if p.line == 0 {
+			ew.write("TIMESTAMP", tab)
+			if p.opts.nodeName {
+				ew.write("NODE", tab)
+			}
 			ew.write(
-				"TIMESTAMP", tab,
-				"NODE", tab,
 				"TYPE", tab,
 				"DETAILS", newline,
 			)
-			p.line++
 		}
-		ew.write(fmtTimestamp(
-			r.GetTime()), tab,
-			r.GetNodeName(), tab,
+		ew.write(fmtTimestamp(r.GetTime()), tab)
+		if p.opts.nodeName {
+			ew.write(r.GetNodeName(), tab)
+		}
+		ew.write(
 			e.GetType(), tab,
 			getAgentEventDetails(e), newline,
 		)
@@ -504,16 +525,21 @@ func (p *Printer) WriteProtoAgentEvent(r *observerpb.GetFlowsResponse) error {
 			return fmt.Errorf("failed to write out agent event: %v", ew.err)
 		}
 	case CompactOutput:
+		var node string
+
+		if p.opts.nodeName {
+			node = fmt.Sprintf(" [%s]", r.GetNodeName())
+		}
 		_, err := fmt.Fprintf(p.opts.w,
-			"%s [%s]: %s\n",
+			"%s%s: %s\n",
 			fmtTimestamp(r.GetTime()),
-			r.GetNodeName(),
+			node,
 			e.GetType())
 		if err != nil {
 			return fmt.Errorf("failed to write out agent event: %v", err)
 		}
 	}
-
+	p.line++
 	return nil
 }
```
</details>

File: pkg/printer/printer.go
Comment (glibsm) [inline pkg/printer/printer.go:463]: for performance reasons, instead of joining strings multiple times, a string builder or `bytes.Buffer` should be used

Comment (tklauser) [reply pkg/printer/printer.go]: Done, thanks.

File: pkg/printer/printer.go
Comment (kAworu) [inline pkg/printer/printer.go:488]: We're not honoring `p.opts.nodeName` here:
```
% ./hubble --server localhost:4245 observe --debug --print-node-name -o json -t agent | jq
{
  "type": "IPCACHE_UPSERTED",
  "ipcache_update": {
    "cidr": "0.0.0.0/0",
    "identity": 2,
    "old_identity": 2
  }
}
...
```
Interestingly, we don't have this issue with flows as we have [a `node_name` field in the Flow protobuf message](https://github.com/cilium/cilium/blob/44c122d930bafffd95805c3b91a3831ceecf9bfd/api/v1/flow/flow.proto#L46-L47). We also don't have this issue with the jsonpb output because there is a `node_name` field in the GetFlowsResponse message.

I think it's fine as-is though, but maybe we want to document that the `--print-node-name` flag has no effect on `json` and `jsonpb` output (which is the status quo).

**Review by kAworu — APPROVED** [review]
:+1: LGTM!

**Comment by tklauser** [comment]
Waiting for #443 to be merged before merging this.

File: pkg/printer/printer.go
Comment (Rolinh) [inline pkg/printer/printer.go:388]: Please, use [`net.JoinHostPort`](https://golang.org/pkg/net/#JoinHostPort) so that IPv6 addresses are properly supported.

File: pkg/printer/printer.go
Comment (Rolinh) [reply pkg/printer/printer.go:488]: The idea was that the node name is always present in `json` and `jsonpb` output.

File: pkg/printer/printer.go
Comment (Rolinh) [inline pkg/printer/printer.go:424]: It feels a bit counterintuitive to have the pod name before the namespace. Would you mind swapping the order?

**Review by Rolinh — CHANGES_REQUESTED** [review]
Looks pretty good overall! The only blocker for me is the comment about not using `net.JoinHostPort` to format the address.

File: pkg/printer/printer.go
Comment (tklauser) [reply pkg/printer/printer.go:388]: Thanks for catching this.

File: pkg/printer/printer.go
Comment (tklauser) [reply pkg/printer/printer.go:424]: Agree. I assume same goes for `pb.AgentEventType_SERVICE_UPSERTED` case below where we want the service name before the namespace?

**Review by Rolinh — APPROVED** [review]

**Comment by tklauser** [comment]
FWIW, I've amended the `cilium/cilium` bump (first commit) to latest master as of cilium/cilium@002bd9c37c4d in order to include the `flow.proto` changes from cilium/cilium#14305.

File: pkg/printer/printer.go
Comment (tklauser) [reply pkg/printer/printer.go:388]: 🤦 I addressed this locally but forget to amend and push before I clicked the merge button. Sorry about that 😞 will send a follow-up PR.

Comment (tklauser) [reply pkg/printer/printer.go:388]: Addressed in #454

File: pkg/printer/printer.go
Comment (tklauser) [reply pkg/printer/printer.go]: Forgot to amend this change as well :facepalm: now addressed in #454

**Pull request merged**
